File: go-algorithm/leetcode/problem525.go
package main

// 525. Contiguous Array
// https://leetcode.com/problems/contiguous-array/

func findMaxLength(nums []int) int {
	if nums[0] == 0 {
		nums[0] = -1
	}
	ans := 0
	for i := 1; i < len(nums); i++ {
		if nums[i] == 1 {
			nums[i] = nums[i-1] + 1
		} else {
			nums[i] = nums[i-1] - 1
		}
		if nums[i] == 0 && i > ans {
			ans = i + 1
		}
	}
	store := make(map[int]int, 0)
	for i := 0; i < len(nums); i++ {
		if val, ok := store[nums[i]]; !ok {
			store[nums[i]] = i
		} else if i-val > ans {
			ans = i - val
		}
	}
	return ans
}

File: java-algo/src/main/java/com/leetcode/Problem525.java
package com.leetcode;

import java.util.HashMap;
import java.util.Map;

/**
 * 525. Contiguous Array
 * https://leetcode.com/problems/contiguous-array/
 */
public class Problem525 {
    public int findMaxLength(int[] nums) {
        Map<Integer, Integer> diff = new HashMap<>();
        diff.put(0, -1);
        int[] cnt = new int[2];
        int ans = 0;
        for (int i = 0; i < nums.length; i++) {
            cnt[nums[i]]++;
            int d = cnt[1] - cnt[0];
            if (diff.containsKey(d)) {
                ans = Math.max(ans, i - diff.get(d));
            } else {
                diff.put(d, i);
            }
        }
        return ans;
    }
}
